### DISCUSSION TITLE
Looking for themes under directories listed in XDG_DATA_DIRS

### DISCUSSION THREAD

**Discussion by romildo** [tsujan/Kvantum#658] [General]
It seems that Kvantum looks for themes under a single prefix directory (`/usr/share` on most systems).

In [NixOS](https://nixos.org/) (a GNU/Linux distribution), the entire operating system, including the kernel, applications, system packages and configuration files, are built by the Nix package manager. Nix stores all packages in isolation from each other, in its own directory under /nix/store.

Therefore Kvantum themes installed by individual packages in NixOS are not put in a common prefix, like in more traditional systems, and Kvantum is not able to find them.

For Kvantum to find themes installed by other packages, it would have to be able to look at other directories other than `/usr/share`. A good solution would be to look at all directories listed in the `XDG_DATA_DIRS` environment variable.

So I would like to request this change in Kvantum, if the authors find it reasonable.

Related to this discussion one can find the following issue reported at the NixOS repository: https://github.com/NixOS/nixpkgs/issues/82769

**Comment by tsujan** [comment]
> It seems that Kvantum looks for themes under a single prefix directory

I don't know of any other Qt style engine that's more flexible than Kvantum in that regard. Please read `Theme-Making.pdf` → `Theme Installation Paths` and also `Dark variants of themes`.

**Reply by romildo** [reply]
After reading the recommended sections of the documentation, I see that (system wide) themes installed at `$DATADIR/Kvantum` can be found by Kvantum. But unless I am wrong, `$DATADIR` (commonly `/usr/share`) should be a unique common parent directory to all of them.

The problem we face with Nix is that there is not a data directory common to several packages. Each package is installed isolated from the others. For instance, on my system the Kvantum theme files for the packages `arc-kde-theme` and `adapta-kde-theme` are installed in the following paths:
``` console
$ ls -l /nix/store/av2f7w10yr4iai7dpv5r56i8cn6na52n-arc-kde-theme-20180614/share/Kvantum
total 12
dr-xr-xr-x 2 root root 4096 Dec 31  1969 Arc
dr-xr-xr-x 2 root root 4096 Dec 31  1969 ArcDark
dr-xr-xr-x 2 root root 4096 Dec 31  1969 ArcDarker

$ ls -l /nix/store/qd4z9rii9125xmy9p0cb70n71dkysk7m-adapta-kde-theme-20180828/share/Kvantum
total 8
dr-xr-xr-x 2 root root 4096 Dec 31  1969 Adapta
dr-xr-xr-x 2 root root 4096 Dec 31  1969 AdaptaNokto
```

The [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) defines the `$XDG_DATA_DIRS` environment variable:
> There is a set of preference ordered base directories relative to which data files should be searched. This set of directories is defined by the environment variable `$XDG_DATA_DIRS`.

If Kvantum could look for system wide themes in all directories listed in `$XDG_DATA_DIRS` (instead of just `$DATADIR`), it would suffice to have this variable to include the data directories used by individual packages. In the example above:
``` console
XDG_DATA_DIRS="$XDG_DATA_DIRS:/nix/store/av2f7w10yr4iai7dpv5r56i8cn6na52n-arc-kde-theme-20180614/share:/nix/store/qd4z9rii9125xmy9p0cb70n71dkysk7m-adapta-kde-theme-20180828/share"
```

**Reply by tsujan** [reply]
> $DATADIR (commonly /usr/share) should be a unique common parent directory to all of them.

Yes, but `XDG_DATA_DIRS` is also a set of specific directories. It seems to me very impractical to add 20 extra directories to `XDG_DATA_DIRS` for 20 extra Kvantum themes!

**Comment by sagikazarmark** [comment]
I have a working implementation here: https://github.com/sagikazarmark/Kvantum/commit/01989083f9ee75a013c2654e760efd0a1dea4a68

Theme path detection should probably be refactored a bit, because this change made it incredibly complex, not to mention that it required changes in several different parts of the code.
